src/bin/pgcopydb/dump_restore.c

```diff
@@ -567,6 +570,167 @@ copydb_target_finalize_schema(CopyDataSpec *specs)
+	char *name = item->restoreListName;
+
+	bool skip = false;
+
+	/*
+	 * Always skip DATABASE objets as pgcopydb does not create the target
+	 * database.
+	 */
+	if (item->desc == ARCHIVE_TAG_DATABASE)
+	{
+		skip = true;
+		log_notice("Skipping DATABASE \"%s\"", name);
+	}
+
+	/*
+	 * Skip COMMENT ON EXTENSION when either of the option
+	 * --skip-extensions or --skip-ext-comment has been used.
+	 */
+	if ((specs->skipExtensions ||
+		 specs->skipCommentOnExtension) &&
+		item->isCompositeTag &&
+		item->tagKind == ARCHIVE_TAG_KIND_COMMENT &&
+		item->tagType == ARCHIVE_TAG_TYPE_EXTENSION)
+	{
+		skip = true;
+		log_notice("Skipping COMMENT ON EXTENSION \"%s\"", name);
+	}
+
+	if (!skip && catOid == PG_NAMESPACE_OID)
+	{
+		bool exists = false;
+
+		if (!copydb_schema_already_exists(specs, oid, &exists))
+		{
+			log_error("Failed to check if restore name \"%s\" "
+					  "already exists",
+					  name);
+			return false;
+		}
+
+		if (exists)
+		{
+			skip = true;
+
+			log_notice("Skipping already existing dumpId %d: %s %u %s",
+					   item->dumpId,
+					   item->description,
+					   item->objectOid,
+					   item->restoreListName);
+		}
+	}
+
+	if (!skip && copydb_objectid_has_been_processed_already(specs, item))
+	{
+		skip = true;
+
+		log_notice("Skipping already processed dumpId %d: %s %u %s",
+				   item->dumpId,
+				   item->description,
+				   item->objectOid,
+				   item->restoreListName);
+	}
+
+	/*
+	 * For SEQUENCE catalog entries, we want to limit the scope of the hash
+	 * table search to the OID, and bypass searching by restore name. We
+	 * only use the restore name for the SEQUENCE OWNED BY statements.
+	 *
+	 * This also allows complex filtering of sequences that are owned by
+	 * table a and used as a default value in table b, where table a has
+	 * been filtered-out from pgcopydb scope of operations, but not table
+	 * b.
+	 */
+	if (item->desc == ARCHIVE_TAG_SEQUENCE)
+	{
+		name = NULL;
+	}
+
+	/*
+	 * There could be a case where the materalized view is included in the
+	 * dump, but the refresh is filtered out using [exclude-table-data].
+	 * In this case, we want to skip the refresh.
+	 */
+	if (!skip &&
+		item->desc == ARCHIVE_TAG_REFRESH_MATERIALIZED_VIEW &&
+		copydb_matview_refresh_is_filtered_out(specs, oid))
+	{
+		skip = true;
+
+		log_notice("Skipping materialized view refresh dumpId %d: %s %u %s",
+				   item->dumpId,
+				   item->description,
+				   item->objectOid,
+				   item->restoreListName);
+	}
+
+	if (!skip && copydb_objectid_is_filtered_out(specs, oid, name))
+	{
+		skip = true;
+
+		log_notice("Skipping filtered-out dumpId %d: %s %u %u %s",
+				   item->dumpId,
+				   item->description,
+				   item->catalogOid,
+				   item->objectOid,
+				   item->restoreListName);
+	}
+
+	/*
+	 * Since printfPQExpBuffer generates a NULL-terminated string,
+	 * we have the ability to use the same buffer repeatedly.
+	 */
+	printfPQExpBuffer(line, "%s%d; %u %u %s %s\n",
```

Comment:
It seems to me we should still call `PQExpBufferBroken(line)` each time we call `printPQExpBuffer` and report an OOM error in case the buffer is broken. If not, then we should add a comment explaining why that's not necessary.